### PR TITLE
[Feat] Add `id` to `title` of `Graph`, `Table`, `AgGrid` and `Container`

### PR DESCRIPTION
**Pull request by huong-li-nguyen**
## Description
- Aligned with @antonymilne that we will provide an `id` to the title components only for now. We might extend further to the header and footer.
- This enables dynamically changing the title based on an action, which I quite commonly use for data viz (see example)

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/examples/scratch_dev/app.py
Comment (petar-qb) [reply vizro-core/examples/scratch_dev/app.py:20]: I agree that the custom action docs example with the timeout is not much useful. We added this one in the docs just to show how custom actions could be created without specifying any `inputs` or `outputs`. I also agree that we have to update actions examples in the docs, but I'd wait until actions v2 is merged because it brings many advantages (one of these enhancements is that your example could filter graph data_frame and additionally update its title).

**Review by petar-qb — APPROVED** [review]
I'm fine with adding id to the title. I think we should consider adding more ids to other components as well, so our users can easily target it with the custom css/js/actions.

**Comment by huong-li-nguyen** [comment]
> I'm fine with adding id to the title. I think we should consider adding more ids to other components as well, so our users can easily target it with the custom css/js/actions.

Did you ever encounter a subcomponent you wanted to target but you couldn't because an `id` was not provided? We can definitely add it in, so would be curious what use case you had! We can definitely add it in if it's a common one.

**Comment by nadijagraca** [comment]
> > I'm fine with adding id to the title. I think we should consider adding more ids to other components as well, so our users can easily target it with the custom css/js/actions.
> 
> Did you ever encounter a subcomponent you wanted to target but you couldn't because an `id` was not provided? We can definitely add it in, so would be curious what use case you had! We can definitely add it in if it's a common one. So far our rule was that the "reactive" part of a component gets the `self.id`. Whether the subcomponents also need to be reactive is something we've decided case-by-case, so it's very useful if you can track your use cases where you wanted to target some subcomponent, but couldn't :)

I needed to target vm.Conteiner title to be changed based on what parameter is selected on yesterdays CNX example. I guess Graph title could do, but Container title would work better for visual cohesion.

File: vizro-core/examples/scratch_dev/app.py
Comment (huong-li-nguyen) [reply vizro-core/examples/scratch_dev/app.py:20]: Do you guys have a rough time estimate when the docs for actions V2 will be updated? If this takes another 2-3 weeks then I would prefer updating it now already, given that it's just a quick replacement of example and we have the code for it already?

Comment (petar-qb) [reply vizro-core/examples/scratch_dev/app.py:20]: Optimistically, the action v2 documents will be merged by the end of October, so I agree we should replace this "time.sleep" action example in the docs. 

However, I don't think that this example from the scratch folder should replace the "time.sleep" action example as this one uses action's "inputs"/"outputs" and we already have some docs examples for that case. Also, this scratch example can confuse our users as it doesn't filer the graph (but the vm.Filter model is used). 

If you find any better example that doesn't use action's inputs/outputs, I'm fine to replace the dummy "time.sleep" one. And, I'm also okay with completely deleting this example from the docs if we can't find any better example.

Comment (huong-li-nguyen) [reply vizro-core/examples/scratch_dev/app.py:20]: I definitely see the point on the `vm.Filter`. I forgot that it replaces the action instead of appending.. 🤔 

On the improved example - why do we even want to show an example without input/output? Is this something we commonly do? In the Dash docs example, they also show a basic callback with input/output which I find more intuitive. While there is a current example with input/output, I find that one a bit more advanced. I was looking for a replicated example similar to the Dash docs where `XXX.value` affects `XXX.children`

**Review by antonymilne — APPROVED** [review]
Just a few comments but generally looks good!

This reminds me of something we should try and achieve eventually with actions v2: it's very counterintuitive for a user trying to figure out what is the right component id and property to target. e.g. in order to change a title of a graph with id `my_graph` you need to target `my_graph_title.children`. Ideally we would simplify this to something like this scheme:
* each model field corresponds to a target property that gets translated automatically into the right Dash component. User target the model field rather than the Dash property. So you'd target `my_graph.title` in order to change the value you originally supplied to the `title` field and things would "just work"
* each model has a default target property (for `Graph` this would be `figure`) so that if you specify `my_graph` with no properties at all it works as a shortcut notation

@petar-qb do we have this tracked somewhere?

File: vizro-core/examples/scratch_dev/app.py
Comment (petar-qb) [reply vizro-core/examples/scratch_dev/app.py:20]: @huong-li-nguyen @antonymilne
If you think it fits, we can also adjust the simple "time.sleep" example to:

```py
@capture("action")
def mock_upload_downloaded_file_to_s3(mock_s3_url: str):
    """Mocked custom action that represents."""
    sleep(2)
   # Uploading downloaded file to s3
   # ...

...

        vm.Button(
            text="Export data",
            actions=[
                vm.Action(function=export_data()),
                vm.Action(
                    function=mock_upload_downloaded_file_to_s3(mock_s3_url="s3://my-bucket-name/<KEY>.csv")
                ),
                vm.Action(function=export_data(file_format="xlsx")),
            ]
        )
```

**Comment by petar-qb** [comment]
> Just a few comments but generally looks good!
> This reminds me of something we should try and achieve eventually with actions v2: it's very counterintuitive for a user trying to figure out what is the right component id and property to target. e.g. in order to change a title of a graph with id `my_graph` you need to target `my_graph_title.children`. Ideally we would simplify this to something like this scheme:
> * each model field corresponds to a target property that gets translated automatically into the right Dash component. User target the model field rather than the Dash property. So you'd target `my_graph.title` in order to change the value you originally supplied to the `title` field and things would "just work"
> * each model has a default target property (for `Graph` this would be `figure`) so that if you specify `my_graph` with no properties at all it works as a shortcut notation

> @petar-qb do we have this tracked somewhere?

Yep, this is the final goal we want to achieve probably with something called "actions_info". This is discussed a lot on a calls and private slack messages but I haven't found the ticket for that (as there are no tickets for most of the Backend V2 stuff that exists on the internal Miro board)

File: vizro-core/src/vizro/models/_components/container.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/container.py:60]: I saw that too. We have been switching around often. This uses the [BEM convention](https://css-tricks.com/bem-101/), which at some point we wanted to implement because it's the recommended CSS convention. It was also the one Pruthvi mentioned we should use.

However, because neither dash, bootstrap nor mantine use BEM, at some point we said that we should just stick with dashes. So we have this mix now 😄 I can tidy this up in a separate PR, but just know we would consciously deviate from the recommended best practices to stay consistent with the other dash-libraries.

File: vizro-core/examples/scratch_dev/app.py
Comment (huong-li-nguyen) [reply vizro-core/examples/scratch_dev/app.py:27]: Sounds good! 

I've (1) left the time.sleep example as is, (2) used the updated example from @antonymilne above and (3) simplified the existing input/output example with `clickData`. 

I think the configuration for that looked more complex than it needed to be. For example, it had a control, filter_interaction, and some more advanced layouts, which I stripped off now because I think the example was supposed to only show "how to create a custom action that shows the clickData of a chart in a Card." 

And for that purpose, I don't think we need the filter, the second chart, the chart interaction, etc.. But let me know if I should revert that change: https://github.com/mckinsey/vizro/pull/714/commits/501228821ca70a3a430a55cbfdc880ecf861f5a2

File: vizro-core/src/vizro/models/_components/container.py
Comment (antonymilne) [reply vizro-core/src/vizro/models/_components/container.py]: Hmmm yes, I thought this was a left over bit of BEM. I think we should stay consistent with the other Dash libraries and tidy this up to go but definitely not a big deal and not for this PR. In fact, might be a good candidate for a Grace Hopper ticket? I'm trying to get some very simple examples like this: https://github.com/mckinsey/vizro/issues/721.

Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/container.py]: Yes that works!

File: vizro-core/examples/scratch_dev/app.py
Comment (antonymilne) [reply vizro-core/examples/scratch_dev/app.py:27]: Yeah, sounds good, let's do that 👍

File: vizro-core/src/vizro/models/_components/container.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/container.py]: https://github.com/mckinsey/vizro/issues/722

**Pull request merged**
